File: system/src/de/uka/ilkd/key/gui/ExampleChooser.java
// This file is part of KeY - Integrated Deductive Software Design
//
// Copyright (C) 2001-2011 Universitaet Karlsruhe (TH), Germany
//                         Universitaet Koblenz-Landau, Germany
//                         Chalmers University of Technology, Sweden
// Copyright (C) 2011-2014 Karlsruhe Institute of Technology, Germany
//                         Technical University Darmstadt, Germany
//                         Chalmers University of Technology, Sweden
//
// The KeY system is protected by the GNU General
// Public License. See LICENSE.TXT for details.
//

package de.uka.ilkd.key.gui;

import java.awt.BorderLayout;
import java.awt.Dimension;
import java.awt.FlowLayout;
import java.awt.Font;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.LinkedList;
import java.util.List;
import java.util.Properties;

import javax.swing.BoxLayout;
import javax.swing.JButton;
import javax.swing.JComponent;
import javax.swing.JDialog;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JSplitPane;
import javax.swing.JTabbedPane;
import javax.swing.JTextArea;
import javax.swing.JTree;
import javax.swing.KeyStroke;
import javax.swing.border.TitledBorder;
import javax.swing.event.TreeSelectionEvent;
import javax.swing.event.TreeSelectionListener;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeModel;
import javax.swing.tree.TreePath;
import javax.swing.tree.TreeSelectionModel;


public final class ExampleChooser extends JDialog {
    /**
     * This path is also accessed by the Eclipse integration of KeY
     * to find the right examples.
     */
    public static final String EXAMPLES_PATH = "examples";

    private static final long serialVersionUID = -4405666868752394532L;

    /**
     * This constant is accessed by the eclipse based projects.
     */
    public static final String KEY_FILE_NAME = "project.key";

    private static final String PROOF_FILE_NAME = "project.proof";

    /**
     * This class is a singleton class and this its only instance.
     */
    private static ExampleChooser instance;

    private final JTree exampleList;
    private final JButton loadButton;
    private final JButton loadProofButton;
    private final JButton cancelButton;
    private JTabbedPane tabPane;

    /**
     * The result value of the dialog. <code>null</code> if nothing to be loaded
     */
    private File fileToLoad = null;

    /**
     * The currently selected example. <code>null</code> if none selected
     */
    private Example selectedExample;

    /**
     * This class wraps a {@link File} and has a special {@link #toString()} method
     * only using the short file name w/o path.
     *
     * Used for displaying files in the examples list w/o prefix
     */
    public static class Example {
        /**
         * The default category under which examples range if they do not
         * have {@link #KEY_PATH} set.
         */
        private static final String DEFAULT_CATEGORY_PATH = "Unsorted";

        /**
         * The {@link Properties} key to specify the path in the tree.
         */
        private static final String KEY_PATH = "example.path";

        /**
         * The {@link Properties} key to specify the name of the example.
         * Directory name if left open.
         */
        private static final String KEY_NAME = "example.name";

        /**
         * The {@link Properties} key to specify the file for the example.
         * KEY_FILE_NAME by default
         */
        private static final String KEY_FILE = "example.file";

        /**
         * The {@link Properties} key to specify the proof file in the tree.
         * May be left open
         */
        private static final String KEY_PROOF_FILE = "example.proofFile";

        /**
         * The {@link Properties} key to specify the path in the tree.
         * Prefix to specify additional files to load. Append 1, 2, 3, ...
         */
        private static final String ADDITIONAL_FILE_PREFIX = "example.additionalFile.";

        /**
         * The {@link Properties} key to specify the path in the tree.
         * Prefix to specify export files which are not shown as tabs in the example wizard but are extracted to Java projects in the Eclipse integration.
         * Append 1, 2, 3, ...
         */
        private static final String EXPORT_FILE_PREFIX = "example.exportFile.";

        private final File exampleFile;
        private final File directory;
        private final String description;
        private final Properties properties;

        public Example(File file) throws IOException {
            this.exampleFile = file;
            this.directory = file.getParentFile();
            this.properties = new Properties();
            StringBuilder sb = new StringBuilder();
            extractDescription(file, sb, properties);
            this.description = sb.toString();
        }

        public File getDirectory() {
           return directory;
        }

        public File getProofFile() {
           return new File(directory, properties.getProperty(KEY_PROOF_FILE, PROOF_FILE_NAME));
        }

        public File getObligationFile() {
            return new File(directory, properties.getProperty(KEY_FILE, KEY_FILE_NAME));
        }

        public String getName() {
            return properties.getProperty(KEY_NAME, directory.getName());
        }

        public String getDescription() {
            return description;
        }

        public File getExampleFile() {
            return exampleFile;
        }

        public List<File> getAdditionalFiles() {
            ArrayList<File> result = new ArrayList<File>();
            int i = 1;
            while(properties.containsKey(ADDITIONAL_FILE_PREFIX + i)) {
                result.add(new File(directory, properties.getProperty(ADDITIONAL_FILE_PREFIX + i)));
                i++;
            }
            return result;
        }

        public List<File> getExportFiles() {
            ArrayList<File> result = new ArrayList<File>();
            int i = 1;
            while(properties.containsKey(EXPORT_FILE_PREFIX + i)) {
                result.add(new File(directory, properties.getProperty(EXPORT_FILE_PREFIX + i)));
                i++;
            }
            return result;
        }

        public String[] getPath() {
            return properties.getProperty(KEY_PATH, DEFAULT_CATEGORY_PATH).split("/");
        }

        @Override
        public String toString() {
            return getName();
        }

        public void addToTreeModel(DefaultTreeModel model) {
            DefaultMutableTreeNode node =
                    findChild((DefaultMutableTreeNode) model.getRoot(), getPath(), 0);
            node.add(new DefaultMutableTreeNode(this));
        }

        private DefaultMutableTreeNode findChild(DefaultMutableTreeNode root, String[] path, int from) {
            if(from == path.length) {
                return root;
            }
            Enumeration<?> en = root.children();
            while(en.hasMoreElements()) {
                DefaultMutableTreeNode node = (DefaultMutableTreeNode) en.nextElement();
                if(node.getUserObject().equals(path[from])) {
                    return findChild(node, path, from + 1);
                }
            }
            // not found ==> add new
            DefaultMutableTreeNode node = new DefaultMutableTreeNode(path[from]);
            root.add(node);
            return findChild(node, path, from + 1);
        }

        public boolean hasProof() {
            return properties.containsKey(KEY_PROOF_FILE);
        }

    }

    //-------------------------------------------------------------------------
    //constructors
    //-------------------------------------------------------------------------

    private ExampleChooser(File examplesDir) {
	super(MainWindow.getInstance(), "Load Example", true);
	assert examplesDir != null;
	assert examplesDir.isDirectory();

	//create list panel
	final JPanel listPanel = new JPanel();
	listPanel.setLayout(new BorderLayout());
	getContentPane().add(listPanel);

	//create example list
	final DefaultTreeModel model = new DefaultTreeModel(new DefaultMutableTreeNode());
	List<Example> examples = listExamples(examplesDir);
	for (Example example : examples) {
	   example.addToTreeModel(model);
	}

	exampleList = new JTree();
	exampleList.setModel(model);
	exampleList.getSelectionModel().setSelectionMode(TreeSelectionModel.SINGLE_TREE_SELECTION);
	exampleList.setRootVisible(false);
	exampleList.addTreeSelectionListener(
	        new TreeSelectionListener() {
	            @Override
	            public void valueChanged(TreeSelectionEvent e) {
	                updateDescription();
	            }
		});
	exampleList.addMouseListener(new MouseAdapter() {
        @Override
	    public void mouseClicked(MouseEvent e){
		if(e.getClickCount() == 2){
		    loadButton.doClick();
		}
	    }
	});
	final JScrollPane exampleScrollPane = new JScrollPane(exampleList);
	exampleScrollPane.setBorder(new TitledBorder("Examples"));

	//create description label
	tabPane = new JTabbedPane(JTabbedPane.TOP);

	JSplitPane split = new JSplitPane(JSplitPane.HORIZONTAL_SPLIT);
	split.add(exampleScrollPane);
	split.add(tabPane);
	split.setDividerLocation(300);
	listPanel.add(split, BorderLayout.CENTER);

	//create button panel
	final JPanel buttonPanel = new JPanel();
	buttonPanel.setLayout(new FlowLayout(FlowLayout.RIGHT, 5, 5));
	final Dimension buttonDim = new Dimension(140, 27);
        buttonPanel.setMaximumSize(new Dimension(Integer.MAX_VALUE,
                                                 (int)buttonDim.getHeight()
                                                     + 10));
	getContentPane().add(buttonPanel);

	//create "load" button
	loadButton = new JButton("Load Example");
	loadButton.addActionListener(new ActionListener() {
        @Override
	    public void actionPerformed(ActionEvent e) {
	        assert selectedExample != null;
		fileToLoad = selectedExample.getObligationFile();
		setVisible(false);
	    }
	});
	buttonPanel.add(loadButton);
	getRootPane().setDefaultButton(loadButton);

	//create "load proof" button
        loadProofButton = new JButton("Load Proof");
        loadProofButton.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                assert selectedExample != null;
                assert selectedExample.hasProof();
                fileToLoad = selectedExample.getProofFile();
                setVisible(false);
            }
        });
        buttonPanel.add(loadProofButton);

	//create "cancel" button
	cancelButton = new JButton("Cancel");
	cancelButton.addActionListener(new ActionListener() {
        @Override
	    public void actionPerformed(ActionEvent e) {
	        fileToLoad = null;
		setVisible(false);
	    }
	});
	buttonPanel.add(cancelButton);
        ActionListener escapeListener = new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent event) {
                if(event.getActionCommand().equals("ESC")) {
                    cancelButton.doClick();
                }
            }
        };
        cancelButton.registerKeyboardAction(
                            escapeListener,
                            "ESC",
                            KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0),
                            JComponent.WHEN_IN_FOCUSED_WINDOW);

        // select first example
        DefaultMutableTreeNode firstLeaf = ((DefaultMutableTreeNode)model.getRoot()).getFirstLeaf();
        TreePath pathToFirstLeaf = new TreePath(firstLeaf.getPath());
        exampleList.getSelectionModel().setSelectionPath(pathToFirstLeaf);
        exampleList.makeVisible(pathToFirstLeaf);

	// show
        getContentPane().setLayout(new BoxLayout(getContentPane(),
                                                 BoxLayout.Y_AXIS));
	setSize(800,400);
	setLocationRelativeTo(MainWindow.getInstance());
    }



    //-------------------------------------------------------------------------
    //internal methods
    //-------------------------------------------------------------------------

    private static File lookForExamples() {
        // greatly simplified version without parent path lookup.
        return new File(System.getProperty("key.home"), EXAMPLES_PATH);
    }

    private static String fileAsString(File f) {
        try {
            byte[] buffer = new byte[(int) f.length()];
            FileInputStream fis = new FileInputStream(f);
            fis.read(buffer);
            fis.close();
            return new String(buffer);
        } catch (IOException e) {
            return "<Error reading file: " + f + ">";
        }
    }

    private static StringBuilder extractDescription(File file, StringBuilder sb, Properties properties)
            throws IOException {
        BufferedReader r = null;
        try {
            r = new BufferedReader(new FileReader(file));
            String line;
            boolean emptyLineSeen = false;
            while((line = r.readLine()) != null) {
                if(emptyLineSeen) {
                    sb.append(line).append("\n");
                } else {
                    String trimmed = line.trim();
                    if(trimmed.length() == 0) {
                        emptyLineSeen = true;
                    } else if(trimmed.startsWith("#")) {
                        // ignore
                    } else {
                        String[] entry = trimmed.split(" *[:=] *", 2);
                        if(entry.length > 1) {
                            properties.put(entry[0], entry[1]);
                        }
                    }
                }
            }
        } finally {
            if(r != null) {
                try { r.close(); }
                catch(IOException ex) { };
            }
        }
        return sb;
    }



    private void updateDescription() {

        TreePath selectionPath = exampleList.getSelectionModel().getSelectionPath();
        if(selectionPath == null) {
            return;
        }

        DefaultMutableTreeNode node =
                (DefaultMutableTreeNode) selectionPath.getLastPathComponent();
        Object nodeObj = node.getUserObject();
        tabPane.removeAll();

        if (nodeObj instanceof Example) {
            Example example = (Example) nodeObj;

            if(example != selectedExample) {
                addTab(example.getDescription().toString(), "Description", true);
                final String fileAsString = fileAsString(example.getObligationFile());
                final int p = fileAsString.lastIndexOf("\\problem");
                if (p >= 0) {
                    addTab(fileAsString.substring(p), "Proof Obligation", false);
                }
                for (File file : example.getAdditionalFiles()) {
                    addTab(fileAsString(file), file.getName(), false);
                }
                loadButton.setEnabled(true);
                loadProofButton.setEnabled(example.hasProof());
                selectedExample = example;
            }
        } else {
            selectedExample = null;
            loadButton.setEnabled(false);
            loadProofButton.setEnabled(false);
        }
    }

    //-------------------------------------------------------------------------
    //public interface
    //-------------------------------------------------------------------------

    private void addTab(String string, String name, boolean wrap) {
        JTextArea area = new JTextArea();
        area.setText(string);
        area.setFont(new Font(Font.MONOSPACED, Font.PLAIN, area.getFont().getSize()));
        area.setCaretPosition(0);
        area.setEditable(false);
        area.setWrapStyleWord(true);
        area.setLineWrap(wrap);
        tabPane.add(new JScrollPane(area), name);
    }

    /**
     * Shows the dialog, using the passed examples directory. If null is passed,
     * tries to find examples directory on its own.
     */
    public static File showInstance(String examplesDirString) {
        // get examples directory
        File examplesDir;
        if(examplesDirString == null) {
            examplesDir = lookForExamples();
        } else {
            examplesDir = new File(examplesDirString);
        }

        if (examplesDir == null || !examplesDir.isDirectory()) {
            JOptionPane.showMessageDialog(MainWindow.getInstance(),
                    "The examples directory cannot be found.\n"+
                    "Please install them at " +
                            (examplesDirString == null ? System.getProperty("key.home") +"/" : examplesDirString),
                    "Error loading examples",
                    JOptionPane.ERROR_MESSAGE);
            return null;
        }

	//show dialog
	if(instance == null) {
	    instance = new ExampleChooser(examplesDir);
	}
	instance.setVisible(true);

	//return result
	final File result = instance.fileToLoad;
	return result;
    }

    /**
     * Lists all examples in the given directory.
     * This method is also accessed by the eclipse based projects.
     * @param examplesDir The examples directory to list examples in.
     * @return The found examples.
     */
    public static List<Example> listExamples(File examplesDir) {
        List<Example> result = new LinkedList<Example>();

        String line;
        BufferedReader br = null;
        try {
            br = new BufferedReader(
                    new FileReader(new File(new File(examplesDir, "index"), "samplesIndex.txt")));
            while((line = br.readLine()) != null) {
                line = line.trim();
                if(line.startsWith("#") || line.length() == 0) {
                    continue;
                }
                File f = new File(examplesDir, line);
                try {
                    result.add(new Example(f));
                } catch (IOException e) {
                    System.err.println("Cannot parse example " +  f + "; ignoring it.");
                    e.printStackTrace();
                }
            }
        } catch (IOException e) {
            System.err.println("Error while reading samples");
            e.printStackTrace();
        } finally {
            try {
                if(br != null) {
                    br.close();
                }
            } catch (IOException e) {
                System.err.println("Error while reading samples");
                e.printStackTrace();
            }
        }

       return result;
    }
}

File: key.fxui/src/main/kotlin/view/ExampleChooser.kt
/* This file is part of KeY - https://key-project.org
 * KeY is licensed under the GNU General Public License Version 2
 * SPDX-License-Identifier: GPL-2.0-only */
package view

import javafx.beans.binding.Bindings
import javafx.scene.control.*
import javafx.scene.text.Font
import org.key_project.util.java.IOUtil
import org.slf4j.LoggerFactory
import tornadofx.*
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Paths
import java.util.*
import javax.swing.tree.TreePath


data class TreeValue(
    val label: String,
    val example: Example? = null,
    val children: MutableList<TreeValue> = arrayListOf()
)

/**
 * Dialog to choose an example to load.
 * May be opened automatically when KeY starts.
 */
class ExampleChooser(val examplesDir: File) : Dialog<File>() {
    val selectedExampleProperty = objectProperty<Example?>(null)
    var selectedExample by selectedExampleProperty

    private var exampleList: TreeView<TreeValue> by singleAssign()
    private var tabPane: TabPane by singleAssign()

    private val root = splitpane {
        titledpane("Examples") {
            isAnimated=false
            isCollapsible=false
            val examples = listExamples(examplesDir)
            val root = TreeValue("/")
            examples.forEach { addToRoot(it, root, it.path.iterator()) }
            exampleList = treeview(TreeItem(root)) {
                isShowRoot = false
                selectionModel.selectionMode = SelectionMode.SINGLE
                onUserSelect {
                    selectedExample = it.example
                }
                cellFormat { text = it.label }
                populate { it.value.children }
            }
        }
        tabPane = tabpane {

        }
        setDividerPositions(.3)
    }

    init {
        dialogPane.content = root
        val load = ButtonType("Load", ButtonBar.ButtonData.APPLY)
        val loadProof = ButtonType("Load Proof", ButtonBar.ButtonData.FINISH)
        val close = ButtonType.CANCEL
        dialogPane.buttonTypes.setAll(load, loadProof, close)


        val hasSelectedExample = selectedExampleProperty.isNotNull
        (dialogPane.lookupButton(load) as Button).enableWhen(hasSelectedExample)
        (dialogPane.lookupButton(loadProof) as Button).enableWhen(hasSelectedExample)

        setResultConverter {
            when (it.buttonData) {
                ButtonBar.ButtonData.APPLY -> {
                    close()
                    selectedExample?.obligationFile
                }
                else -> null
            }
        }




        // create example list

        /*loadButton = JButton("Load Example")
        loadButton.addActionListener(ActionListener { e: ActionEvent? ->
            if (selectedExample == null) {
                throw RuntimeException("No example selected")
            }
            fileToLoad = selectedExample!!.obligationFile
            setVisible(false)
        })
        buttonPanel.add(loadButton)
        getRootPane().setDefaultButton(loadButton)

        // create "load proof" button
        loadProofButton = JButton("Load Proof")
        loadProofButton.addActionListener(ActionListener { e: ActionEvent? ->
            checkNotNull(selectedExample) { "No example selected" }
            check(selectedExample!!.hasProof()) { "Selected example has no proof." }
            fileToLoad = selectedExample!!.proofFile
            setVisible(false)
        })
        buttonPanel.add(loadProofButton)

        // create "cancel" button
        cancelButton = JButton("Cancel")
        cancelButton.addActionListener(ActionListener { e: ActionEvent? ->
            fileToLoad = null
            setVisible(false)
        })
        buttonPanel.add(cancelButton)
        GuiUtilities.attachClickOnEscListener(cancelButton)

        // select first example
        val firstLeaf: DefaultMutableTreeNode = (model.getRoot() as DefaultMutableTreeNode).getFirstLeaf()
        val pathToFirstLeaf = TreePath(firstLeaf.getPath())
        exampleList.selectionModel.setSelectionPath(pathToFirstLeaf)
        exampleList.makeVisible(pathToFirstLeaf)

        // show
        getContentPane().setLayout(BoxLayout(getContentPane(), BoxLayout.Y_AXIS))
         */
        dialogPane.setPrefSize(800.0, 400.0)

        selectedExampleProperty.onChange {
            if (it != null) updateDescription(it) else {
                tabPane.tabs.clear()
                /*selectedExample = null
            loadButton.setEnabled(false)
            loadProofButton.setEnabled(false)
             */
            }
        }
    }

    private tailrec fun addToRoot(example: Example, node: TreeValue, path: Iterator<String>) {
        if (path.hasNext()) {
            val p = path.next()
            val c = node.children.find { it.label == p }
            val n = c ?: TreeValue(p, null).also { node.children.add(it) }
            addToRoot(example, n, path)
        } else {
            node.children.add(TreeValue(example.name, example))
        }
    }


    private fun updateDescription(nodeObj: Example) {
        tabPane.tabs.clear()
        addTab(nodeObj.description, "Description", true)
        val fileAsString = nodeObj.obligationFile.readText()
        val p = fileAsString.lastIndexOf("\\problem")
        if (p >= 0) {
            addTab(fileAsString.substring(p), "Proof Obligation", false)
        }
        for (file in nodeObj.additionalFiles) {
            addTab(fileAsString(file), file.getName(), false)
        }
        //loadButton.setEnabled(true)
        //loadProofButton.setEnabled(nodeObj.hasProof())
        selectedExample = nodeObj
    }

    private fun addTab(string: String, name: String, wrap: Boolean) {
        val area = TextArea(string)
        area.font = Font("monospace", area.font.size)
        area.positionCaret(0)
        area.isEditable = false
        area.isWrapText = wrap
        val tab = Tab(name, area)
        tab.isClosable = false
        tabPane.tabs.add(tab)
    }

    companion object {
        /**
         * This path is also accessed by the Eclipse integration of KeY to find the right examples.
         */
        const val EXAMPLES_PATH = "examples"

        /**
         * This constant is accessed by the eclipse based projects.
         */
        const val KEY_FILE_NAME = "project.key"
        const val PROOF_FILE_NAME = "project.proof"

        /**
         * Java property name to specify a custom key example folder.
         */
        const val KEY_EXAMPLE_DIR = "key.examples.dir"
        val LOGGER = LoggerFactory.getLogger(ExampleChooser::class.java)

        /**
         * This class is a singleton class and this its only instance.
         */
        private var instance: ExampleChooser? = null

        // -------------------------------------------------------------------------
        // internal methods
        // -------------------------------------------------------------------------
        fun lookForExamples(): File {
            // weigl: using java properties: -Dkey.examples.dir="..."
            if (System.getProperty(KEY_EXAMPLE_DIR) != null) {
                return File(System.getProperty(KEY_EXAMPLE_DIR))
            }

            return File("key.ui/examples").absoluteFile

        }

        private fun fileAsString(f: File): String {
            return try {
                IOUtil.readFrom(f)
            } catch (e: IOException) {
                LOGGER.error("Could not read file '{}'", f, e)
                "<Error reading file: $f>"
            }
        }

        fun extractDescription(
            file: File, sb: StringBuilder,
            properties: Properties
        ): StringBuilder {
            val lines = file.readLines()
            var emptyLineSeen = false
            for (line in lines) {
                if (emptyLineSeen) {
                    sb.append(line).append("\n")
                } else {
                    val trimmed = line.trim { it <= ' ' }
                    if (trimmed.isEmpty()) {
                        emptyLineSeen = true
                    } else if (trimmed.startsWith("#")) {
                        // ignore
                    } else {
                        val entry = trimmed.split(" *[:=] *".toRegex(), limit = 2).toTypedArray()
                        if (entry.size > 1) {
                            properties[entry[0]] = entry[1]
                        }
                    }
                }
            }
            return sb
        }

        /**
         * Shows the dialog, using the passed examples directory. If null is passed, tries to find
         * examples directory on its own.
         */
        fun showInstance(examplesDirString: String? = null): File? {
            // get examples directory
            val examplesDir: File = examplesDirString?.let { File(examplesDirString) } ?: lookForExamples()
            if (!examplesDir.isDirectory()) {
                alert(
                    Alert.AlertType.ERROR, "Error loading examples",
                    """
                        The examples directory cannot be found.
                        Please install them at ${
                        examplesDirString ?: IOUtil.getProjectRoot(ExampleChooser::class.java).toString()
                    }/
                        """
                )
                return null
            }
            // show dialog
            val dialog = instance ?: ExampleChooser(examplesDir)
            instance = dialog
            return dialog.showAndWait().orElse(null)
        }

        /**
         * Lists all examples in the given directory. This method is also accessed by the eclipse based
         * projects.
         *
         * @param examplesDir The examples directory to list examples in.
         * @return The found examples.
         */
        fun listExamples(examplesDir: File): List<Example> {
            val result: MutableList<Example> = LinkedList()
            val index = examplesDir / "index" / "samplesIndex.txt"

            try {
                for (read in index.readLines()) {
                    val line = read.trim { it <= ' ' }
                    if (line.startsWith("#") || line.isEmpty()) {
                        continue
                    }
                    val f = examplesDir / line
                    try {
                        result.add(Example(f))
                    } catch (e: IOException) {
                        LOGGER.warn("Cannot parse example {}; ignoring it.", f, e)
                    }
                }
            } catch (e: IOException) {
                LOGGER.warn("Error while reading samples", e)
            }
            return result
        }
    }
}

private operator fun File.div(s: String) = File(this, s)

/**
 * This class wraps a [File] and has a special [.toString] method only using the
 * short file name w/o path.
 *
 *
 * Used for displaying files in the examples list w/o prefix
 */
class Example(val exampleFile: File) {
    val directory: File = exampleFile.getParentFile()
    private val properties: Properties = Properties()
    val description: String

    init {
        val sb = StringBuilder()
        ExampleChooser.extractDescription(exampleFile, sb, properties)
        description = sb.toString()
    }

    val proofFile: File
        get() = File(directory, properties.getProperty(KEY_PROOF_FILE, ExampleChooser.PROOF_FILE_NAME))
    val obligationFile: File
        get() = File(directory, properties.getProperty(KEY_FILE, ExampleChooser.KEY_FILE_NAME))
    val name: String
        get() = properties.getProperty(KEY_NAME, directory.getName())
    val additionalFiles: List<File>
        get() {
            val result = ArrayList<File>()
            var i = 1
            while (properties.containsKey(ADDITIONAL_FILE_PREFIX + i)) {
                result.add(File(directory, properties.getProperty(ADDITIONAL_FILE_PREFIX + i)))
                i++
            }
            return result
        }
    val exportFiles: List<File>
        get() {
            val result = ArrayList<File>()
            var i = 1
            while (properties.containsKey(EXPORT_FILE_PREFIX + i)) {
                result.add(File(directory, properties.getProperty(EXPORT_FILE_PREFIX + i)))
                i++
            }
            return result
        }
    val path: Array<String>
        get() = properties.getProperty(KEY_PATH, DEFAULT_CATEGORY_PATH).split("/".toRegex())
            .dropLastWhile { it.isEmpty() }
            .toTypedArray()

    companion object {
        /**
         * The default category under which examples range if they do not have [.KEY_PATH]
         * set.
         */
        private const val DEFAULT_CATEGORY_PATH = "Unsorted"

        /**
         * The [Properties] key to specify the path in the tree.
         */
        private const val KEY_PATH = "example.path"

        /**
         * The [Properties] key to specify the name of the example. Directory name if left
         * open.
         */
        private const val KEY_NAME = "example.name"

        /**
         * The [Properties] key to specify the file for the example. KEY_FILE_NAME by default
         */
        private const val KEY_FILE = "example.file"

        /**
         * The [Properties] key to specify the proof file in the tree. May be left open
         */
        private const val KEY_PROOF_FILE = "example.proofFile"

        /**
         * The [Properties] key to specify the path in the tree. Prefix to specify additional
         * files to load. Append 1, 2, 3, ...
         */
        private const val ADDITIONAL_FILE_PREFIX = "example.additionalFile."

        /**
         * The [Properties] key to specify the path in the tree. Prefix to specify export
         * files which are not shown as tabs in the example wizard but are extracted to Java
         * projects in the Eclipse integration. Append 1, 2, 3, ...
         */
        private const val EXPORT_FILE_PREFIX = "example.exportFile."
    }
}
